### ISSUE TITLE
npm audit security issue with handlebars 4.0.x

### ISSUE THREAD

**Issue by cholz** [pillarjs/hbs#176]
NPM Packages built with "audit" will fail, due a handlebars security issue.
Updating the handlebars dependency to 4.3.1 should solve the issue.

```
                       === npm audit security report ===


                                 Manual Review
             Some vulnerabilities require your attention to resolve

          Visit https://go.npm.me/audit-guide for additional guidance


  High            Prototype Pollution

  Package         handlebars

  Patched in      >=4.3.0

  Dependency of   hbs

  Path            hbs > handlebars

  More info       https://nodesecurity.io/advisories/1164

found 1 high severity vulnerability in 1765970 scanned packages
  1 vulnerability requires manual review. See the full report for details.
```

**Comment by dougwilson** [comment]
Thank you for the report. Looks like it was just published yesterday. I will look at what changes need to be done to upgrade; I seem to recall there was an incompatibility moving to the 4.1.x series I had on my backlog to resolve, but I guess I need to do this today: +1:

**Comment by Naktibalda** [comment]
I hope that this issue can be fixed in a minor or patch release, bumping version number to 5.0.0 would require other libraries to update their `hbs` version constraint.

**Comment by mcandre** [comment]
Unfortunately, this upgrade is going to be a lot harder than it needs to be.

* handlebars 4.3.0 breaks semver.
* hbs appears to be testing dependencies with istanbul?
* validation is much too strict:

```console
$ npm test

> hbs@4.0.4 test /Users/tkmamhf/go/src/github.com/pillarjs/hbs
> istanbul cover node_modules/mocha/bin/_mocha


  ․ express 3.x index: 41ms
  ․ express 3.x partials: 6ms
  ․ express 3.x html extension: 12ms
  1) express 3.x syntax error

  3 passing (2s)
  1 failing

  1) express 3.x syntax error:
     AssertionError [ERR_ASSERTION]: Input A expected to strictly equal input B:
+ expected - actual

- 'Error'
+ 'Error: /Users/tkmamhf/go/src/github.com/pillarjs/hbs/test/3.x/views/syntax-error.hbs: Parse error on line 1:'
      at /Users/tkmamhf/go/src/github.com/pillarjs/hbs/test/3.x/app.js:246:12
      at Test._assertFunction (/Users/tkmamhf/go/src/github.com/pillarjs/hbs/node_modules/supertest/lib/test.js:247:11)
      at Test.assert (/Users/tkmamhf/go/src/github.com/pillarjs/hbs/node_modules/supertest/lib/test.js:148:18)
      at Server.assert (/Users/tkmamhf/go/src/github.com/pillarjs/hbs/node_modules/supertest/lib/test.js:127:12)
      at Object.onceWrapper (events.js:273:13)
      at Server.emit (events.js:182:13)
      at emitCloseNT (net.js:1650:8)
      at process._tickCallback (internal/process/next_tick.js:63:19)



=============================================================================
Writing coverage object [/Users/tkmamhf/go/src/github.com/pillarjs/hbs/coverage/coverage.json]
Writing coverage reports at [/Users/tkmamhf/go/src/github.com/pillarjs/hbs/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 65.71% ( 115/175 )
Branches     : 47.76% ( 32/67 )
Functions    : 69.23% ( 27/39 )
Lines        : 65.71% ( 115/175 )
================================================================================
npm ERR! Test failed.  See above for more details.
```

**Comment by dougwilson** [comment]
Yes, I agree. I would like to figure out one of three paths is all (maybe there are more)

1. What can we do to upgrade to latest handlebars but keep this a 4.x release so folks get it automatically with semver ranges.
2. Maybe even if we can backport the fix to the 4.0.x branch of handlebars (this was done with previous patches) thus we can bump the patch version.
3. Last case, document the breaking changes and migration guide and release as 5.0 and people will need to migrate to address the issue.

Those are my thoughts, at least, in order of own own preferences.

**Comment by dougwilson** [comment]
I have traced through the one test failure and created a bug in the handlebars project for it: https://github.com/wycats/handlebars.js/issues/1562

Our test suite is not really great at coverage, so I am just going through all the changes between the 4.0.14 handlebars and 4.3.1 currently. Even though I filed the bug above, I believe I found a work-around, so it won't end up as a blocker thankfully.

**Comment by dougwilson** [comment]
Just a heads-up for those looking here, there is yet another security issue handlebars is working to fix: https://github.com/wycats/handlebars.js/issues/1563 . I continue to work towards the upgrade here, which will make us ready for when that new handlebars version will drop.

**Comment by dougwilson** [comment]
A new version of this module, `hbs`, will be published with the updated `handlebars` dependency within the next 8 hours.

**Comment by dougwilson** [comment]
Ok, handlebars 4.3.3 is published now, so that will be the upgrade target. Also, AFAICT there isn't anything breaking when used through this module, so this will be a patch release for `hbs`.

**Comment by dougwilson** [comment]
Published as 4.0.5 with handlebars 4.3.3
